package/MDAnalysis/coordinates/H5MD.py

```diff
@@ -726,6 +777,635 @@ def __setstate__(self, state):
+            except AttributeError:
+                errmsg = "Input obj is neither an AtomGroup or Universe"
+                raise TypeError(errmsg) from None
+
+        if ts.n_atoms != self.n_atoms:
+            raise IOError("H5MDWriter: Timestep does not have"
+                          " the correct number of atoms")
+
+        # This should only be called once when first timestep is read.
+        if self.h5md_file is None:
+            self._determine_units(ag)
+            self._open_file()
+            self._initialize_hdf5_datasets(ts)
+
+        return self._write_next_timestep(ts)
+
+    def _determine_units(self, ag):
+        """determine which units the file will be written with
+
+        By default, it fills the :attr:`self.units` dictionary by copying the
+        units dictionary of the parent file. Because H5MD files have no
+        standard unit restrictions, users may pass a kwarg in ``(timeunit,
+        lengthunit, velocityunit, forceunit)`` to the writer so long as
+        MDAnalysis has a conversion factor for it (:exc:`ValueError` raised if
+        it does not). These custom unit arguments must be in
+        `MDAnalysis notation`_. If custom units are supplied from the user,
+        :attr`self.units[unit]` is replaced with the corresponding
+        ``unit`` argument.
+
+        """
+
+        # use units of parent file as native units
+        try:
+            # ag == AtomGroup
+            self.units = ag.universe.trajectory.units.copy()
+        except AttributeError:
+            # ag == Universe
+            self.units = ag.trajectory.units.copy()
+
+        # set user input units
+        for key, value in self._new_units.items():
+            if value is not None:
+                if value not in self._unit_translation_dict[key]:
+                    raise ValueError(f"{value} is not a unit recognizable by"
+                                     " MDAnalyis. Allowed units are:"
+                                     f" {self._unit_translation_dict.keys()}"
+                                     " For more information on units, see"
+                                     " `MDAnalyis units`_.")
+                else:
+                    self.units[key] = self._new_units[key]
+
+        if self.convert_units:
+            # check if all units are None
+            if not any(self.units.values()):
+                raise ValueError("The trajectory has no units, but "
+                                 "``convert_units`` is set to ``True`` by "
+                                 "default in MDAnalysis. To write the file "
+                                 "with no units, set ``convert_units=False``.")
+
+    def _open_file(self):
+        """Opens file with `H5PY`_ library and fills in metadata from kwargs.
+
+        :attr:`self.h5md_file` becomes file handle that links to root level.
+
+        """
+
+        if self._comm is not None:  # pragma: no cover
+            # ValueError raised if communicator is passed
```

Comment:
say where he ValueError is raised, currently a bit confusing because the raise is not here